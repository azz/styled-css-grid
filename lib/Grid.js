import styled from "styled-components";
import PropTypes from "prop-types";

const autoRows = ({ minRowHeight = "20px" }) => `minmax(${minRowHeight}, auto)`;

const frGetter = value =>
  typeof value === "number" ? `repeat(${value}, 1fr)` : value;

const gap = ({ gap = "8px" }) => gap;

const flow = ({ flow = "row" }) => flow;

const formatAreas = areas => areas.map(area => `"${area}"`).join(" ");

const Grid = styled.div`
  display: grid;
  grid-auto-flow: ${flow};
  grid-auto-rows: ${autoRows};
  ${({ rows }) => rows && `grid-template-rows: ${frGetter(rows)}`};
  grid-template-columns: ${({ columns = 12 }) => frGetter(columns)};
  grid-gap: ${gap};
  ${({ areas }) => areas && `grid-template-areas: ${formatAreas(areas)}`};
  ${({ justifyContent }) =>
    justifyContent && `justify-content: ${justifyContent}`};
  ${({ alignContent }) => alignContent && `align-content: ${alignContent}`};
`;

Grid.propTypes = {
  className: PropTypes.string,
  columns: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  gap: PropTypes.string,
  minRowHeight: PropTypes.string,
  flow: PropTypes.string,
  rows: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  areas: PropTypes.arrayOf(PropTypes.string),
  justifyContent: PropTypes.string,
  alignContent: PropTypes.string
};

export default Grid;
