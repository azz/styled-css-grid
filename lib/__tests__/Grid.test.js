import React from "react";
import renderer from "react-test-renderer";
import "jest-styled-components";

import Grid from "../Grid";

describe("<Grid />", () => {
  test("matches snapshot with default args", () => {
    const tree = renderer.create(<Grid />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("'columns' as number prop sets 'grid-template-columns' to repeat()", () => {
    const tree = renderer.create(<Grid columns={7} />).toJSON();
    expect(tree).toHaveStyleRule("grid-template-columns", "repeat(7,1fr)");
  });

  test("'columns' as string prop sets 'grid-template-columns'", () => {
    const tree = renderer
      .create(<Grid columns="repeat(auto-fit, minmax(100px, 1fr))" />)
      .toJSON();
    expect(tree).toHaveStyleRule(
      "grid-template-columns",
      "repeat(auto-fit,minmax(100px,1fr))"
    );
  });

  test("'rows' as number prop sets 'grid-template-rows' to repeat()", () => {
    const tree = renderer.create(<Grid rows={7} />).toJSON();
    expect(tree).toHaveStyleRule("grid-template-rows", "repeat(7,1fr)");
  });

  test("'rows' as string prop sets 'grid-template-rows'", () => {
    const tree = renderer.create(<Grid rows="1fr 2fr 1fr" />).toJSON();
    expect(tree).toHaveStyleRule("grid-template-rows", "1fr 2fr 1fr");
  });

  test("'gap' prop sets 'grid-gap'", () => {
    const tree = renderer.create(<Grid gap="7px" />).toJSON();
    expect(tree).toHaveStyleRule("grid-gap", "7px 7px");
  });

  test("'minRowHeight' prop sets 'grid-auto-rows'", () => {
    const tree = renderer.create(<Grid minRowHeight="7px" />).toJSON();
    expect(tree).toHaveStyleRule("grid-auto-rows", "minmax(7px,auto)");
  });

  test("'flow' prop sets 'grid-auto-flow'", () => {
    const tree = renderer.create(<Grid flow="column" />).toJSON();
    expect(tree).toHaveStyleRule("grid-auto-flow", "column");
  });

  test("'areas' array prop sets 'grid-template-areas'", () => {
    const tree = renderer.create(<Grid areas={["a a", "b b"]} />).toJSON();
    expect(tree).toHaveStyleRule("grid-template-areas", '"a a" "b b"');
  });

  test("'alignContent' prop sets 'align-content'", () => {
    const tree = renderer.create(<Grid alignContent="end" />).toJSON();
    expect(tree).toHaveStyleRule("align-content", "end");
  });

  test("'justifyContent' prop sets 'justify-content'", () => {
    const tree = renderer.create(<Grid justifyContent="end" />).toJSON();
    expect(tree).toHaveStyleRule("justify-content", "end");
  });
});
