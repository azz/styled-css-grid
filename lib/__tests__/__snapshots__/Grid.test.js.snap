// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> matches snapshot with default args 1`] = `
.c0 {
  display: grid;
  grid-auto-flow: row;
  grid-auto-rows: minmax(20px,auto);
  grid-template-columns: repeat(12,1fr);
  grid-gap: 8px 8px;
}

<div
  className="c0"
/>
`;
